### PR TITLE
chore(enriching): move enrichment tables into new crate

### PR DESCRIPTION
**Pull request by StephenWakely**
Ref #8815 , specifically [this comment](https://github.com/timberio/vector/issues/8815#issuecomment-905696548).

This moves the Enrichment Table functionality into a new crate under `lib/enrichment`.  

VRL no longer has any knowledge of enrichment tables. The data is passod through between Vector and the `get_enrichment_table_record` via a new `external_context` array. `external_context` is a `Vec<Box<dyn Any>>`, so can be used to pass any data between the client program and a VRL function.

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

File: lib/vrl/compiler/src/state.rs [hunk 1]
```diff
@@ -93,6 +86,7 @@ impl Compiler {
     /// Roll back the compiler state to a previously stored snapshot.
     pub(crate) fn rollback(&mut self) {
         if let Some(snapshot) = self.snapshot.take() {
+            // TODO Copy the data..
```
Comment (JeanMertz) [inline lib/vrl/compiler/src/state.rs:89]: ☝️

File: lib/vrl/core/src/lib.rs [hunk 2]
```diff
@@ -2,19 +2,20 @@ pub mod prelude;
...
 pub use diagnostic;
 pub use runtime::{Runtime, RuntimeResult, Terminate};
+use std::any::Any;
 
 /// Compile a given source into the final [`Program`].
 pub fn compile(
     source: &str,
-    enrichment_tables: Box<dyn enrichment::TableSetup>,
     fns: &[Box<dyn Function>],
+    external_context: Vec<Box<dyn Any>>,
```
Comment (JeanMertz) [inline lib/vrl/core/src/lib.rs:15]: Do we have a reason to make this a `Vec`? I think just `Option<Box<dyn Any>>` would suffice for the current use-case, wouldn't it?

If so, I'd prefer that, we can always change it we have a need for more external context to be added.

**Review by JeanMertz — CHANGES_REQUESTED** [review]
This is definitely a much preferable solution 👌.

One change request, and a TODO comment left, but otherwise this LGTM.

File: lib/vector-core/src/transform/config.rs [hunk 3]
```diff
@@ -1,6 +1,5 @@
 use crate::config::GlobalOptions;
 #[cfg(feature = "vrl")]
```
Comment (blt) [inline lib/vector-core/src/transform/config.rs:2]: You also need to remove this feature flag.

File: lib/vector-core/Cargo.toml [hunk 4]
```diff
@@ -16,6 +15,7 @@ chrono = { version = "0.4.19", default-features = false, features = ["serde"] }
 db-key = { version = "0.0.5", default-features = false, optional = true }
 derivative = { version = "2.2.0", default-features = false }
 dyn-clone = { version = "1.0.4", default-features = false }
+enrichment = { path = "../enrichment" }
```
Comment (blt) [inline lib/vector-core/Cargo.toml:18]: I could be wrong, but `enrichment` is not actually a dependency of vector-core any longer?

**Review by blt — CHANGES_REQUESTED** [review]

File: lib/vrl/compiler/src/state.rs (hunk 1 above)
Comment (StephenWakely) [reply lib/vrl/compiler/src/state.rs:89]: Good catch, thanks.

File: lib/vrl/core/src/lib.rs (hunk 2 above)
Comment (StephenWakely) [reply lib/vrl/core/src/lib.rs:15]: `Option` would work for now.

File: lib/vector-core/Cargo.toml (hunk 4 above)
Comment (StephenWakely) [reply lib/vector-core/Cargo.toml:18]: It is used, but only when the `vrl` feature is enabled. I'll update.

**Review by JeanMertz — APPROVED** [review]
Sweet! Three nits, but otherwise LGTM

**Review by blt — APPROVED** [review]
Nice work.

**Pull request merged**